src/div_ops.cxx

```diff
@@ -951,6 +929,353 @@ const Field3D Div_a_Grad_perp_upwind(const Field3D& a, const Field3D& f) {
   return result;
 }
 
+// Div ( a Grad_perp(f) ) -- ∇⊥ ( a ⋅ ∇⊥ f) --  Vorticity
+// Includes nonorthogonal X-Y and X-Z metric corrections
+//
+Field3D Div_a_Grad_perp_nonorthog(const Field3D& a, const Field3D& f) {
+  ASSERT2(a.getLocation() == f.getLocation());
+
+  Mesh* mesh = a.getMesh();
+
+  Coordinates* coord = f.getCoordinates();
+
+  // Y and Z fluxes require Y derivatives
+
+  // Fields containing values along the magnetic field
+  Field3D fup(mesh), fdown(mesh);
+  Field3D aup(mesh), adown(mesh);
+
+  Field3D g23up(mesh), g23down(mesh);
+  Field3D g_23up(mesh), g_23down(mesh);
+  Field3D g12up(mesh), g12down(mesh);
+  Field3D g_12up(mesh), g_12down(mesh);
+  Field3D Jup(mesh), Jdown(mesh);
+  Field3D dyup(mesh), dydown(mesh);
+  Field3D dzup(mesh), dzdown(mesh);
+  Field3D Bxyup(mesh), Bxydown(mesh);
+
+  // Values on this y slice (centre).
+  // This is needed because toFieldAligned may modify the field
+  Field3D fc = f;
+  Field3D ac = a;
+
+  Field3D g23c = coord->g23;
+  Field3D g_23c = coord->g_23;
+  Field3D g12c = coord->g12;
+  Field3D g_12c = coord->g_12;
+  Field3D Jc = coord->J;
+  Field3D dxc = coord->dx;
+  Field3D dyc = coord->dy;
+  Field3D dzc = coord->dz;
+  Field3D Bxyc = coord->Bxy;
+
+  // Calculate the X derivative at cell edge (X + 1/2), including in Y guard cells
+  // This is used to calculate Y flux contribution from g21 * d/dx
+  Field3D fddx_xhigh(mesh);
+  fddx_xhigh.allocate();
+  for (int i = mesh->xstart - 1; i <= mesh->xend; i++) {
+    for (int j = mesh->ystart - 1; j <= mesh->yend + 1;
+         j++) { // Note: Including one guard cell
+      for (int k = 0; k < mesh->LocalNz; k++) {
+        fddx_xhigh(i, j, k) = 2. * (f(i + 1, j, k) - f(i, j, k))
+                              / (coord->dx(i, j, k) + coord->dx(i + 1, j, k));
+      }
+    }
+  }
+  Field3D fddx_xhigh_up(mesh), fddx_xhigh_down(mesh);
+
+  // Result of the Y and Z fluxes
+  Field3D yzresult(mesh);
+  yzresult.allocate();
+
+  if (f.hasParallelSlices() && a.hasParallelSlices()) {
+    // Both inputs have yup and ydown
+
+    fup = f.yup();
+    fdown = f.ydown();
+
+    aup = a.yup();
+    adown = a.ydown();
+
+    mesh->communicate(fddx_xhigh);
```

Comment:
Note that this does not work for fci.
for fci the right thing would be to do the above calculation also for the parallel slices.
```suggestion
    if (f.isFci()) {
        throw BoutException("This is not implemented for FCI.\n"
        "The above calculation needs to be done for the parallel slices as well");
    }
    mesh->communicate(fddx_xhigh);
```